File: src/shared-ui/core/dateRanges.js
// Calendar windows for the preset time ranges that are not periods on the wire.
//
// Day / month / total ship with every stats snapshot because the collector scans
// exactly those three windows. "Yesterday" and "this week" are calendar ranges in
// the same sense as a hand-picked custom range, so they resolve through
// /api/usage/range instead: no new wire field, no extra tokscale scan per tick, and
// the same aggregate-plus-detail semantics the range dialog already has.
//
// Bounds are whole local days. The upper bound is the last millisecond of the final
// day rather than the next midnight, because the desktop hosts read `to` as
// "include that whole hour" — midnight-to-midnight would leak the first hour of the
// following day into the total. The Hub rounds `to` down to a calendar day either way.

export const PRESET_RANGE_PERIODS = Object.freeze(['yesterday', 'week']);

/**
 * The day the 本周 window starts on: ISO 8601 Monday, on every surface and in every
 * locale.
 *
 * This is a *measurement* rule, not a display preference. `Intl.Locale#weekInfo` is
 * absent in the Chromium builds both hosts run on, so a locale-driven lookup could only
 * ever return a fallback here — while a client that *does* have CLDR data (Android,
 * Node) would compute a different span and report a different figure for the same 本周
 * label. One window, one number, so a phone and a browser are comparable. Calendar grids
 * and heatmaps are presentation and may start wherever their locale says; nothing that
 * prints a reported total reads this.
 */
export const SCOPE_WEEK_FIRST_DAY_INDEX = 1;

export function isPresetRangePeriod(period) {
  return PRESET_RANGE_PERIODS.includes(String(period || ''));
}

function pad(value) {
  return String(value).padStart(2, '0');
}

export function localDayKey(date) {
  return `${date.getFullYear()}-${pad(date.getMonth() + 1)}-${pad(date.getDate())}`;
}

function startOfDay(date) {
  return new Date(date.getFullYear(), date.getMonth(), date.getDate());
}

// Date's day overflow handles month/year boundaries and DST: adding 1 to the 31st
// rolls over, and a 23- or 25-hour day still lands on the next calendar date.
function addDays(date, days) {
  return new Date(date.getFullYear(), date.getMonth(), date.getDate() + days);
}

function endOfDay(date) {
  return new Date(date.getFullYear(), date.getMonth(), date.getDate(), 23, 59, 59, 999);
}

/**
 * The day the scope bar's 本周 window starts on.
 *
 * A function rather than a bare read of [SCOPE_WEEK_FIRST_DAY_INDEX] so the shape stays
 * familiar to call sites, but it takes no locale: the argument was only ever a fallback
 * lookup here, and letting a host's CLDR data move the window would give the same label a
 * different span on a different device.
 */
export function firstDayOfWeekIndex() {
  return SCOPE_WEEK_FIRST_DAY_INDEX;
}

export function weekStart(date = new Date(), firstDayIndex = SCOPE_WEEK_FIRST_DAY_INDEX) {
  const day = startOfDay(date);
  const back = (day.getDay() - firstDayIndex + 7) % 7;
  return addDays(day, -back);
}

/**
 * @returns {{period: string, startDate: string, endDate: string, from: Date, to: Date}|null}
 *   the inclusive calendar days plus the request bounds, or null for a period that
 *   is not a preset range (those come from the snapshot's `periods`).
 */
export function presetRangeWindow(period, now = new Date()) {
  const name = String(period || '');
  if (!isPresetRangePeriod(name)) return null;
  const today = startOfDay(now);
  // Yesterday is a closed window; the current week always runs up to today.
  const first = name === 'yesterday' ? addDays(today, -1) : weekStart(today);
  const last = name === 'yesterday' ? first : today;
  return {
    period: name,
    startDate: localDayKey(first),
    endDate: localDayKey(last),
    from: first,
    to: endOfDay(last)
  };
}

/** True while a fetched window still answers the same question it was asked for. */
export function presetRangeWindowMatches(expected, active) {
  if (!expected || !active) return false;
  return active.kind === expected.period
    && active.startDate === expected.startDate
    && active.endDate === expected.endDate;
}

/**
 * True when a fetched range answer belongs to the scope tab that is being rendered.
 *
 * A range answer belongs to exactly one selection: the preset that asked for it, or the
 * transient "custom" chip a hand-picked range occupies. Everything else — including the
 * same preset from before midnight — is another tab's number, and rendering it under the
 * current label reports a measurement the user never asked for. This is the predicate the
 * scope chip already uses to pick what to highlight (`rangeKind === 'custom' ? 'custom' :
 * period`), so the highlighted tab and the figure on screen are decided by one rule.
 */
export function isRangeScopeSelection({ period, customRange } = {}) {
  const kind = String(customRange?.kind || '');
  if (!kind) return false;
  if (kind === 'custom') return true;
  return isPresetRangePeriod(kind) && kind === String(period || '');
}

/**
 * Resolve a scope tab to the period it may render.
 *
 * `periods` carries the three windows the collector scans; a preset or picked range is
 * answered by `customPeriod`. When the selection asks for a range the host does not have
 * (loading, failed, or stale after midnight) this returns `null` rather than substituting
 * a snapshot period — the caller renders a loading or retry affordance instead of a
 * number it never obtained.
 */
export function resolveScopePeriod({ period, customRange, customPeriod, periods } = {}) {
  if (customPeriod && isRangeScopeSelection({ period, customRange })) return customPeriod;
  if (isPresetRangePeriod(period)) return null;
  return periods?.[String(period || '')] || null;
}

File: android/app/src/main/java/com/igng/tokenmonitor/android/ui/core/DateRanges.kt
package com.igng.tokenmonitor.android.ui.core

import java.time.DayOfWeek
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter
import java.time.temporal.TemporalAdjusters
import java.util.Locale

/**
 * Calendar windows for the scope presets that are *not* periods on the wire.
 *
 * Day / month / total ship with every stats snapshot because the collector scans
 * exactly those three windows.  "Yesterday" and "this week" are calendar ranges in the
 * same sense as a hand-picked custom range, so they resolve through
 * `GET /api/usage/range` — the same rule the shared web UI uses
 * (`src/shared-ui/core/dateRanges.js`), which is what keeps the two clients reporting
 * the same number for the same tab.  A port that invents its own week boundary would
 * be a second measurement, not a second rendering of one.
 *
 * Bounds are whole local days and inclusive, matching the endpoint's
 * `startDate` / `endDate` / `startHour` / `endHour` vocabulary: the final day is asked
 * for with `endHour = 23`, which is how the desktop host reads "include that whole
 * hour".  `docs/API.md` notes the totals are day-rounded on the Hub side either way.
 */
object DateRanges {
  /** Presets resolved through the range endpoint rather than the snapshot. */
  val PRESET_RANGE_PERIODS = listOf("yesterday", "week")

  /**
   * The 本周 window always starts on ISO 8601 Monday, on every device and in every
   * locale.
   *
   * This is a *measurement* rule, not a display preference: the web scope bar computes
   * the same Monday (`Intl.Locale#weekInfo` is absent in the Chromium builds both hosts
   * run on, and the port pins it deliberately now), so one "本周" label cannot answer two
   * different totals depending on which client asked.  A CLDR-following port was exactly
   * the second measurement the note above warns about — under a US locale its window
   * started a day earlier and the same tab showed a bigger number than the browser.
   *
   * [firstDayOfWeek] stays locale-driven on purpose: the heatmap grid and the date
   * picker's calendar are presentation, and neither one produces a reported figure.
   */
  val SCOPE_WEEK_FIRST_DAY = DayOfWeek.MONDAY

  /**
   * The day a *grid* starts on, from CLDR via [Locale].
   *
   * Display only — see [SCOPE_WEEK_FIRST_DAY] for why the 本周 window ignores this.
   */
  fun firstDayOfWeek(locale: Locale = Locale.getDefault()): DayOfWeek {
    // `WeekFields` is the CLDR-backed accessor; an unsupported locale still yields
    // the ISO minimum, which is the documented fallback rather than a surprise.
    return runCatching {
      java.time.temporal.WeekFields.of(locale).firstDayOfWeek
    }.getOrDefault(SCOPE_WEEK_FIRST_DAY)
  }

  /** The ISO Monday that starts [date]'s week. */
  fun scopeWeekStart(date: LocalDate = LocalDate.now()): LocalDate {
    val back = (date.dayOfWeek.value - SCOPE_WEEK_FIRST_DAY.value + 7) % 7
    return date.minusDays(back.toLong())
  }

  /**
   * @return the inclusive calendar window for a preset name, or null when [period] is
   *   not a preset range (those come from the snapshot's `periods`).
   */
  fun presetRangeWindow(
    period: String?,
    today: LocalDate = LocalDate.now()
  ): PresetRangeWindow? {
    return when (period?.lowercase()) {
      // Yesterday is a closed window; the current week always runs up to today.
      "yesterday" -> PresetRangeWindow(
        period = "yesterday",
        startDate = today.minusDays(1),
        endDate = today.minusDays(1)
      )
      "week" -> PresetRangeWindow(
        period = "week",
        startDate = scopeWeekStart(today),
        endDate = today
      )
      else -> null
    }
  }

  fun isPresetRangePeriod(period: String?): Boolean =
    period?.lowercase() in PRESET_RANGE_PERIODS
}

data class PresetRangeWindow(
  val period: String,
  val startDate: LocalDate,
  val endDate: LocalDate
) {
  /** The `startHour` the range endpoint is called with. */
  val startHour: Int get() = 0

  /** The `endHour` the range endpoint is called with — the last hour of the window. */
  val endHour: Int get() = 23

  /** True while a fetched window still answers the same question it was asked for. */
  fun stillMatches(other: PresetRangeWindow?): Boolean =
    other != null && other.period == period &&
      other.startDate == startDate && other.endDate == endDate

  /**
   * Label for the scope chip.  Both ends are printed even when they are the same day:
   * a single date would hide that "yesterday" is a one-day window, which is the fact
   * a user comparing it against "today" needs.
   */
  fun label(now: LocalDate = LocalDate.now()): String {
    val formatter = DateTimeFormatter.ofPattern("MM-dd")
    return if (startDate == endDate) {
      startDate.format(formatter)
    } else {
      "${startDate.format(formatter)} → ${endDate.format(formatter)}"
    }
  }
}
